sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala

```diff
@@ -18,14 +18,227 @@
+    val schema: StructType = getInsertTableSchema.getOrElse(return None)
+    val schemaWithoutPartitionCols =
+      StructType(schema.fields.dropRight(insert.get.partitionSpec.size))
+    val newDefaultExprs: Seq[Expression] =
+      getDefaultExprs(numQueryOutputs, schemaWithoutPartitionCols)
+    val newAliases: Seq[NamedExpression] =
+      newDefaultExprs.zip(schemaWithoutPartitionCols.fields).map {
+        case (expr, field) => Alias(expr, field.name)()
+      }
+    if (newDefaultExprs.isEmpty) return None
+    Some(project.copy(projectList = project.projectList ++ newAliases))
+  }
+
+  // This is a helper for the addMissingDefaultColumnValues methods above.
+  private def getDefaultExprs(
+      numQueryOutputs: Int, schemaWithoutPartitionCols: StructType): Seq[Expression] = {
+    val remainingFields: Seq[StructField] = schemaWithoutPartitionCols.fields.drop(numQueryOutputs)
+    val numDefaultExprsToAdd: Int = {
+      if (SQLConf.get.useNullsForMissingDefaultColumnValues) {
+        remainingFields.size
+      } else {
+        remainingFields.takeWhile(_.metadata.contains(CURRENT_DEFAULT_COLUMN_METADATA_KEY)).size
+      }
+    }
+    Seq.fill(numDefaultExprsToAdd)(UnresolvedAttribute(CURRENT_DEFAULT_COLUMN_NAME))
+  }
+
+  // Each of the following methods replaces unresolved "DEFAULT" column references with matching
+  // default column values.
+  private def replaceExplicitDefaultColumnValues(
+      analyzer: Analyzer,
+      table: UnresolvedInlineTable): Option[UnresolvedInlineTable] = {
+    assert(insert.isDefined)
+    val schema: StructType = getInsertTableSchema.getOrElse(return None)
+    val schemaWithoutPartitionCols =
+      StructType(schema.fields.dropRight(insert.get.partitionSpec.size))
+    val defaultExprs: Seq[Expression] = schemaWithoutPartitionCols.fields.map {
+      case f if f.metadata.contains(CURRENT_DEFAULT_COLUMN_METADATA_KEY) =>
+        analyze(analyzer, f, "INSERT")
+      case _ => Literal(null)
+    }
+    var replaced = false
+    val newRows: Seq[Seq[Expression]] = {
+      table.rows.map { row: Seq[Expression] =>
+        for {
+          i <- 0 until row.size
+          expr = row(i)
+          defaultExpr = if (i < defaultExprs.size) defaultExprs(i) else Literal(null)
+        } yield expr match {
+          case u: UnresolvedAttribute if isExplicitDefaultColumn(u) =>
+            replaced = true
+            defaultExpr
+          case expr@_ if expr.find{isExplicitDefaultColumn}.isDefined =>
+            // Return a more descriptive error message if the user tries to nest the DEFAULT column
+            // reference inside some other expression, such as DEFAULT + 1 (this is not allowed).
+            throw new AnalysisException(DEFAULTS_IN_EXPRESSIONS_ERROR)
+          case _ => expr
+        }
+      }
+    }
+    if (!replaced) return None
+    Some(table.copy(rows = newRows))
+  }
+
+  private def replaceExplicitDefaultColumnValues(
```

Comment:
There is a lot of duplicated code in this method.
Can we abstract them?

Reply:
Good point, I combined them into one method to share the preparation work.